File: src/test/java/pages/commonPages/BasePage.java
package pages.commonPages;

import com.microsoft.playwright.*;
import com.microsoft.playwright.options.AriaRole;
import com.microsoft.playwright.options.WaitForSelectorState;
import org.junit.Assert;
import utils.Driver;
import utils.GlobalVariables;

import java.time.LocalDate;
import java.time.format.DateTimeFormatter;
import java.util.Arrays;
import java.util.Random;

public class BasePage {

    protected Page page;

    public BasePage()
    {
        page = Driver.get();
    }

    public void clickElement(Locator locator){
        locator.click();
    }

    /**
     * text elementi doğrular.
     * @param value string değeri
     * @param locator locator değeri
     */
    public void verifyTextElement(String value, Locator locator){
        Assert.assertEquals(value, locator.textContent());
    }

    /**
     * trim kullanarak text elementi doğrular.
     * @param value text değeri
     * @param locator locator içindeki text değeri
     */
    public void verifyTextElementUseTrim(String value, Locator locator){
        Assert.assertEquals(value.trim(), locator.textContent().trim());
    }

    /**
     * elementin görünür olmasını kontrol eder.
     * @param locator elementi
     */
    public void verifyIsVisible(Locator locator){
        Assert.assertTrue(locator.isVisible());
    }

    /**
     * gereken değeri başka yerde kullanmak için saklar.
     * @param key string değer kendimiz veririz.
     * @param value locator değerini alır.
     */
    public void addString(String key, String value){
        GlobalVariables.getInstance().addString(key,value);
    }

    /**
     * saklanan string değeri kullanmak için getirir.
     * @param key addString metodunda kendimizin yazdığı değeri karşılar
     * @return değeri döner
     */
    public String getString(String key){
        return GlobalVariables.getInstance().getString(key);
    }

    /**
     * frame içinde işlem yapmaya yarar.
     * @param dialogTitle frame ismi
     * @return frame değerini döner
     */
    public FrameLocator getFrameByDialogTitle(String dialogTitle) {
        return page
                .getByRole(AriaRole.DIALOG, new Page.GetByRoleOptions().setName(dialogTitle))
                .frameLocator("iframe[title='Setur']");
    }

    /**
     * Kendo komponenti kullanıldığı için gereken yerlerde kullanılır.
     * @param frameLocator hangi frame ise
     * @param inputSelector hangi locator ise
     * @param value locatora girilecek değer
     */
    public void setKendoNumericTextBoxValue(FrameLocator frameLocator, String inputSelector, String value) {
        Locator input = frameLocator.locator(inputSelector);
        input.evaluate("(el, val) => {" +
                "  const widget = $(el).data('kendoNumericTextBox');" +
                "  if (widget) {" +
                "    widget.value(val);" +
                "    widget.trigger('change');" +  // veya widget.trigger('input');
                "  }" +
                "}", value);
    }

    public void setKendoNumericTextBoxValue1(Locator input, String value) {
        input.waitFor(new Locator.WaitForOptions().setState(WaitForSelectorState.ATTACHED));

        ElementHandle inputHandle = input.elementHandle();
        if (inputHandle != null) {
            input.page().evaluate("(element, val) => {" +
                    "  const widget = $(element).data('kendoNumericTextBox');" +
                    "  if (widget) {" +
                    "    console.log('Widget bulundu, değer atanıyor');" +
                    "    widget.value(val);" +
                    "    widget.trigger('change');" +
                    "  } else {" +
                    "    console.warn('Kendo NumericTextBox bulunamadı');" +
                    "  }" +
                    "}", Arrays.asList(inputHandle, value));
        } else {
            System.out.println("Element bulunamadı veya görünmüyor.");
        }
    }



    public int generateRandomNumber(){
        Random random = new Random();
        return random.nextInt(10000);
    }
    /**
     * random tarih değeri girer
     * @return tarih değerini döner
     */
    public String generateRandomDate(){
        Random random = new Random();
        int daysToAdd = random.nextInt(30);
        LocalDate randomDate = LocalDate.now().plusDays(daysToAdd);
        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("dd.MM.yyyy");
        return randomDate.format(formatter);
    }
    /**
     * barkod numarası üretir
     * @return string bir barkod no döner
     */
    public String generateBarcodeNumber(){
        Random randomNum = new Random();
        StringBuilder sb = new StringBuilder();
        // İlk rakam 1–9 arasında olmalı (başta 0 olamaz)
        sb.append(randomNum.nextInt(9) + 1);
        // Geri kalan 12 rakam 0–9
        for (int i = 0; i < 12; i++) {
            sb.append(randomNum.nextInt(10));
        }
        return sb.toString();
    }
    /**
     * Gümrük Beyanname No random üretir.
     * @return Gümrük Beyanname No
     */
    public String generateCustomHouseNo() {
        Random random = new Random();
        String part1 = String.format("%08d", random.nextInt(100_000_000));
        String part2 = String.format("%08d", random.nextInt(100_000_000));
        return part1 + "AN" + part2;
    }
    /**
     * pop-up onaylama yapar. (frame içi hariç)
     */
    public void popUpConfirmationProcess(){
        Locator popup = page.locator(".ajs-dialog");
        popup.waitFor(new Locator.WaitForOptions().setState(WaitForSelectorState.VISIBLE));

        Locator okButton = popup.locator(".ajs-button.ajs-ok");
        okButton.click();
    }

    /**
     * pop-up vazgeçme işlemi yapar.
     */
    private void orderCancellationProcess(){
        Locator popup = page.locator(".ajs-dialog");
        popup.waitFor(new Locator.WaitForOptions().setState(WaitForSelectorState.VISIBLE));

        Locator cancelButton = popup.locator(".ajs-button.ajs-cancel");
        cancelButton.click();
    }
}

File: RetailTRUI.Tests/Pages/Common/BasePage.cs
namespace RetailTRUI.Tests.Pages.Common;

/// <summary>
/// Base page class providing common functionality for all page objects
/// Implements core Playwright operations and helper methods
/// </summary>
public abstract class BasePage
{
    private IPage? _pageInstance;
    
    protected IPage Page
    {
        get
        {
            if (_pageInstance != null)
                return _pageInstance;
                
            var driverPage = Driver.Get();
            _pageInstance = driverPage;
            return driverPage;
        }
    }

    protected void ClickElement(ILocator locator)
    {
        locator.ClickAsync().GetAwaiter().GetResult();
    }

    protected async Task ClickElementAsync(ILocator locator)
    {
        await locator.ClickAsync();
    }

    protected void VerifyTextElement(string expectedValue, ILocator locator)
    {
        var actualText = locator.TextContentAsync().GetAwaiter().GetResult();
        actualText.Should().Be(expectedValue);
    }

    protected void VerifyTextElementTrimmed(string expectedValue, ILocator locator)
    {
        var actualText = locator.TextContentAsync().GetAwaiter().GetResult()?.Trim();
        actualText.Should().Be(expectedValue.Trim());
    }

    protected void VerifyIsVisible(ILocator locator)
    {
        locator.IsVisibleAsync().GetAwaiter().GetResult().Should().BeTrue();
    }

    protected async Task<bool> IsVisibleAsync(ILocator locator)
    {
        return await locator.IsVisibleAsync();
    }

    protected void AddString(string key, string value)
    {
        GlobalVariables.Instance.AddString(key, value);
    }

    protected string GetString(string key)
    {
        return GlobalVariables.Instance.GetString(key);
    }

    protected IFrameLocator GetFrameByDialogTitle(string dialogTitle)
    {
        return Page
            .GetByRole(AriaRole.Dialog, new PageGetByRoleOptions { Name = dialogTitle })
            .FrameLocator("iframe[title='Setur']");
    }

    /// <summary>
    /// Sets value for Kendo NumericTextBox component using JavaScript evaluation
    /// </summary>
    protected void SetKendoNumericTextBoxValue(IFrameLocator frameLocator, string inputSelector, string value)
    {
        var input = frameLocator.Locator(inputSelector);
        input.EvaluateAsync($@"
            (el, val) => {{
                const widget = $(el).data('kendoNumericTextBox');
                if (widget) {{
                    widget.value(val);
                    widget.trigger('change');
                }}
            }}", value).GetAwaiter().GetResult();
    }

    /// <summary>
    /// Sets value for Kendo NumericTextBox component on main page
    /// </summary>
    protected async Task SetKendoNumericTextBoxValueAsync(ILocator input, string value)
    {
        await input.WaitForAsync(new LocatorWaitForOptions { State = WaitForSelectorState.Attached });

        var handle = await input.ElementHandleAsync();
        if (handle != null)
        {
            await Page.EvaluateAsync(@"
                (element, val) => {
                    const widget = $(element).data('kendoNumericTextBox');
                    if (widget) {
                        widget.value(val);
                        widget.trigger('change');
                    } else {
                        console.warn('Kendo NumericTextBox not found');
                    }
                }", new object[] { handle, value });
        }
    }

    protected int GenerateRandomNumber()
    {
        return Random.Shared.Next(10000);
    }

    /// <summary>
    /// Generates a random date within the next 30 days
    /// </summary>
    protected string GenerateRandomDate()
    {
        var daysToAdd = Random.Shared.Next(30);
        var randomDate = DateTime.Now.AddDays(daysToAdd);
        return randomDate.ToString("dd.MM.yyyy");
    }

    /// <summary>
    /// Generates a 13-digit barcode number
    /// </summary>
    protected string GenerateBarcodeNumber()
    {
        var sb = new System.Text.StringBuilder();
        // First digit must be 1-9 (cannot start with 0)
        sb.Append(Random.Shared.Next(1, 10));
        // Remaining 12 digits can be 0-9
        for (int i = 0; i < 12; i++)
        {
            sb.Append(Random.Shared.Next(0, 10));
        }
        return sb.ToString();
    }

    /// <summary>
    /// Generates a customs house number in format: 8digits+AN+8digits
    /// </summary>
    protected string GenerateCustomHouseNo()
    {
        var part1 = Random.Shared.Next(0, 100_000_000).ToString("D8");
        var part2 = Random.Shared.Next(0, 100_000_000).ToString("D8");
        return $"{part1}AN{part2}";
    }

    protected async Task WaitForLoadingAsync(int timeoutMs = 5000)
    {
        try
        {
            await Page.WaitForLoadStateAsync(LoadState.NetworkIdle, new PageWaitForLoadStateOptions 
            { 
                Timeout = timeoutMs 
            });
        }
        catch (TimeoutException)
        {
            // Ignore timeout - page may already be loaded
        }
    }

    /// <summary>
    /// Confirms popup dialogs by clicking OK button
    /// </summary>
    protected async Task PopUpConfirmationProcessAsync()
    {
        var okButton = Page.Locator(".ajs-button.ajs-ok");
        await okButton.WaitForAsync(new LocatorWaitForOptions { State = WaitForSelectorState.Visible, Timeout = 5000 });
        await okButton.ClickAsync();
        await Page.WaitForTimeoutAsync(500);
    }

    /// <summary>
    /// Confirms alert popup dialogs
    /// </summary>
    protected async Task ConfirmPopupAsync()
    {
        var okButton = Page.Locator(".ajs-button.ajs-ok");
        await okButton.WaitForAsync(new LocatorWaitForOptions { State = WaitForSelectorState.Visible, Timeout = 3000 });
        await okButton.ClickAsync();
    }
}
